crates/services/upgradable-executor/src/executor.rs

```diff
@@ -604,6 +615,47 @@ where
         }
     }
 
+    /// Attempt to fetch and validate an uploaded WASM module.
+    #[cfg(feature = "wasm-executor")]
+    pub fn validate_uploaded_wasm(
+        &self,
+        wasm_root: &Bytes32,
+    ) -> Result<(), UpgradableError> {
+        self.get_module_by_root_and_validate(*wasm_root).map(|_| ())
+    }
+
+    /// Find an uploaded WASM blob by it's hash and validate it.
+    /// This is a slow operation, and the cached result should be used if possible,
+    /// for instancy by calling `get_module` below.
+    #[cfg(feature = "wasm-executor")]
+    fn get_module_by_root_and_validate(
+        &self,
+        bytecode_root: Bytes32,
+    ) -> Result<wasmtime::Module, UpgradableError> {
+        let view = StructuredStorage::new(
+            self.storage_view_provider
+                .latest_view()
+                .map_err(ExecutorError::from)?,
+        );
+        let uploaded_bytecode = view
+            .storage::<UploadedBytecodes>()
+            .get(&bytecode_root)
+            .map_err(ExecutorError::from)?
+            .ok_or(not_found!(UploadedBytecodes))
+            .map_err(ExecutorError::from)?;
+
+        let fuel_core_types::fuel_vm::UploadedBytecode::Completed(bytecode) =
+            uploaded_bytecode.as_ref()
+        else {
+            return Err(ExecutorError::Other(format!(
```

Comment:
Maybe we want to return a typed error here instead?=)

Reply:
It was like that before as well, so I didn't think it needed changing. Done in 4f1561714